libs/hwes/eventstoredb/projections/custom/custom.go

```diff
@@ -0,0 +1,135 @@
+package custom
+
+import (
+	"context"
+	"fmt"
+	"github.com/EventStore/EventStore-Client-Go/esdb"
+	zlog "github.com/rs/zerolog/log"
+	"hwes"
+	"strings"
+)
+
+const subscriptionGroupName = "custom"
+
+type eventHandler func(ctx context.Context, evt hwes.Event) (error, esdb.Nack_Action)
+
+type CustomProjection struct {
+	es            *esdb.Client
+	eventHandlers map[string]eventHandler
+}
+
+func NewCustomProjection(esdbClient *esdb.Client) *CustomProjection {
+	return &CustomProjection{es: esdbClient, eventHandlers: make(map[string]eventHandler, 0)}
+}
+
+func (p *CustomProjection) RegisterEventListener(eventType string, eventHandler eventHandler) *CustomProjection {
+	if _, found := p.eventHandlers[eventType]; found {
+		zlog.Error().
+			Str("eventType", eventType).
+			Msg("override existing event handler")
+	}
+
+	p.eventHandlers[eventType] = eventHandler
+	return p
+}
+
+func (p *CustomProjection) handleEvent(ctx context.Context, event hwes.Event) (error, esdb.Nack_Action) {
+	eventHandler, found := p.eventHandlers[event.EventType]
+	if !found {
+		return fmt.Errorf("event type '%s' is invalid", event.EventType), esdb.Nack_Unknown
+	}
+	return eventHandler(ctx, event)
+}
+
+func (p *CustomProjection) Subscribe(ctx context.Context) error {
+	log := zlog.Ctx(ctx)
+
+	// Create subscription on EventStoreDB
+	persistentAllSubscriptionOptions := esdb.PersistentAllSubscriptionOptions{}
+	err := p.es.CreatePersistentSubscriptionAll(ctx, subscriptionGroupName, persistentAllSubscriptionOptions)
+	if err != nil {
+		// Ignore subscription already exists error
+		if persistentSubscriptionError, ok := err.(*esdb.PersistentSubscriptionError); !ok || ok && (persistentSubscriptionError.Code != 6) {
+			zlog.Ctx(ctx).Warn().Err(err).Send()
+			return err
```

Comment:
prepend something so we can trace back the error to this line, same for the other error returns
e.g., "projection subscription error:"